core/schema/git.go

```diff
@@ -476,87 +502,96 @@ func (s *gitSchema) withAuthHeader(ctx context.Context, parent *core.GitReposito
 
 type treeArgs struct {
 	DiscardGitDir bool `default:"false"`
+	Depth         int  `default:"1"`
 }
 
 func (s *gitSchema) tree(ctx context.Context, parent dagql.Instance[*core.GitRef], args treeArgs) (inst dagql.Instance[*core.Directory], _ error) {
-	if parent.Self.UseDagOp() && !core.DagOpInContext[core.FSDagOp](ctx) {
-		return DagOpDirectory(ctx, s.srv, parent, args, s.tree, nil)
-	}
-
-	query, ok := s.srv.Root().(dagql.Instance[*core.Query])
-	if !ok {
-		return inst, fmt.Errorf("failed to get root query")
-	}
-	bk, err := query.Self.Buildkit(ctx)
-	if err != nil {
-		return inst, err
-	}
+	if !core.DagOpInContext[core.FSDagOp](ctx) {
+		inst, err := DagOpDirectory(ctx, s.srv, parent, args, s.tree, nil)
+		if err != nil {
+			return inst, err
+		}
 
-	dir, err := parent.Self.Tree(ctx, s.srv, args.DiscardGitDir)
-	if err != nil {
-		return inst, err
-	}
-	inst, err = dagql.NewInstanceForCurrentID(ctx, s.srv, parent, dir)
-	if err != nil {
-		return inst, err
-	}
+		query, ok := s.srv.Root().(dagql.Instance[*core.Query])
+		if !ok {
+			return inst, fmt.Errorf("failed to get root query")
+		}
+		bk, err := query.Self.Buildkit(ctx)
+		if err != nil {
+			return inst, err
+		}
 
-	var dgstInputs []string
+		// use the commit + fully-qualified ref for the base of the content hash
+		var dgstInputs []string
+		var commit string
+		err = s.srv.Select(ctx, parent, &commit, dagql.Selector{Field: "commit"})
+		if err != nil {
+			return inst, err
+		}
+		dgstInputs = append(dgstInputs, commit)
+		var fullref string
+		err = s.srv.Select(ctx, parent, &fullref, dagql.Selector{Field: "ref"})
```

Comment:
why do we need both the commit and the ref mixed in here? Would have expected the commit is all we want

Reply:
Yeah, this is super intentional :tada:

Before, if I checked out a tag + branch that pointed to the exact same commit, I would get a racy git state - either the tag or the branch would be around. It was entirely possible to just be missing one of them, but not reliably (kind of a pain if you ran exec `git show <ref>`, which you'd expect to just work IMO).

Because this PR mimics the behavior of `git checkout <ref>`, we're actually on the checked-out branches now, so this behavior becomes even more important.